src/codegen/compile.ml

```diff
@@ -10315,6 +10315,23 @@ and compile_exp_as env ae sr_out e =
   let sr_in, code = compile_exp_with_hint env ae (Some sr_out) e in
   code ^^ StackRep.adjust env sr_in sr_out
 
+and single_case e (cs : Ir.case list) =
+  match cs, e.note.Note.typ with
+  | [{it={pat={it=TagP (l, _);_}; _}; _}], Type.(Variant [{lab; _}]) -> l = lab
+  | _ -> false
+
+and known_tag_pat p = TagP ("", p)
+
+and simplify_cases e (cs : Ir.case list) =
+  match cs, e.note.Note.typ with
+  (* for a 2-cased variant type, the second comparison can be omitted when the first pattern
+     (with irrefutable subpattern) didn't match, and the patternn types line up *)
+  | [{it={pat={it=TagP (_, ip); note=tp1; _}; _}; _} as c1; {it={pat={it=TagP (_, pat'); note=tp2; _} as pat2; exp}; _} as c2], (Type.(Variant [_; _]) as te)
+       when Ir_utils.is_irrefutable ip
+            && Type.sub tp1 te && Type.sub tp2 te ->
```

Comment:
@crusso either I got this backwards, or there is a bug in the `sub` predicate, as these didn't change a bit 😥

Reply:
Just check both labels are in the type te.

Reply:
did that!